### PR TITLE
Fix lgtm warning

### PR DESCRIPTION
**Pull request by thebevrishot**

### CODE REVIEW

File: src/Ztm.Data.Entity/Contexts/Main/Input.cs [hunk 1]
```diff
@@ -41,5 +41,35 @@ public int CompareTo(Input other)
...
+
+            return CompareTo((Input)other) == 0;
+        }
+
+        public override int GetHashCode()
+        {
+            int hash = 0;
+
+            hash ^= (TransactionHash != null) ? TransactionHash.GetHashCode() : 0;
+            hash ^= (int)Index;
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity/Contexts/Main/Input.cs]: After discussion we decided to turn off this LGTM alert and use `Index.GetHashCode()` instead.

File: src/Ztm.Data.Entity/Contexts/Main/Input.cs [hunk 2]
```diff
@@ -41,5 +41,35 @@ public int CompareTo(Input other)
...
+            hash ^= (OutputHash != null) ? OutputHash.GetHashCode() : 0;
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity/Contexts/Main/Input.cs]: We don't need this lines and below due to this class represent an input on a transaction. So just `TransactionHash` and `Index` should be enough.

File: src/Ztm.Data.Entity/Contexts/Main/Output.cs [hunk 3]
```diff
@@ -41,5 +41,33 @@ public int CompareTo(Output other)
 
             return 0;
         }
+
+        public override bool Equals(Object other)
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity/Contexts/Main/Output.cs]: Same as previous issue.

File: src/Ztm.Data.Entity/Contexts/Main/Output.cs [hunk 4]
```diff
@@ -41,5 +41,33 @@ public int CompareTo(Output other)
...
+        {
+            if (other == null)
+            {
+                return false;
+            }
+
+            if (GetType() != other.GetType())
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity/Contexts/Main/Output.cs]: Same as previous issue.

File: src/Ztm.Data.Entity/Contexts/Main/Output.cs [hunk 5]
```diff
@@ -41,5 +41,33 @@ public int CompareTo(Output other)
...
+
+            return CompareTo((Output)other) == 0;
+        }
+
+        public override int GetHashCode()
+        {
+            int hash = 0;
+
+            hash ^= (TransactionHash != null) ? TransactionHash.GetHashCode() : 0;
+            hash ^= (int)Index;
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity/Contexts/Main/Output.cs]: Same as previous issue.

File: src/Ztm.Data.Entity/Contexts/Main/Output.cs [hunk 6]
```diff
@@ -41,5 +41,33 @@ public int CompareTo(Output other)
...
+            hash ^= (int)Value;
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity/Contexts/Main/Output.cs]: Same as previous issue for this line and below.

File: src/Ztm.Zcoin.Synchronization/AddressWatcher.cs [hunk 7]
```diff
@@ -158,9 +158,8 @@ public class AddressWatcher : ITransactionConfirmationListener
                 }
 
                 // Add debit to address.
-                balances[address] = balances.ContainsKey(address)
-                    ? balances[address] + spend.Value
-                    : spend.Value;
+                Money val;
+                balances[address] = balances.TryGetValue(address, out val) ? val + spend.Value : spend.Value;
```
Comment (ultimaweapon) [inline src/Ztm.Zcoin.Synchronization/AddressWatcher.cs]: We can do something like this `balances.TryGetValue(address, out var val)` and remove the above line.

Comment (ultimaweapon) [inline src/Ztm.Zcoin.Synchronization/AddressWatcher.cs]: IMO better to rename `val` to `previous` will be better.

File: src/Ztm.Data.Entity/Contexts/Main/Input.cs (hunk 2 above)
Comment (thebevrishot) [reply src/Ztm.Data.Entity/Contexts/Main/Input.cs]: Make sense

File: src/Ztm.Zcoin.Synchronization/AddressWatcher.cs (hunk 7 above)
Comment (thebevrishot) [reply src/Ztm.Zcoin.Synchronization/AddressWatcher.cs]: Good to know this sugar syntax

Comment (thebevrishot) [reply src/Ztm.Zcoin.Synchronization/AddressWatcher.cs]: make sense

File: src/Ztm.Data.Entity/Contexts/Main/Block.cs [hunk 8]
```diff
@@ -32,27 +32,22 @@ public int CompareTo(Block other)
...
+            }
 
-                return Hash == other.Hash ? 0 : -1;
+            if (Hash != null)
+            {
+                return Hash.CompareTo(other.Hash);
             }
 
-            return Height - other.Height;
+            return Hash == other.Hash ? 0 : -1;
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity/Contexts/Main/Block.cs]: We need to modify this every time we add more fields to compare. My idea is change the above `if` to:

```cs
if (Hash != null)
{
    return Hash.CompareTo(other.Hash);
}
else if (other.Hash != null)
{
    return -1;
}
```

Then we can change this line to just `return 0;`.

**Pull request merged**